Constant propagation of nondet-symbol [blocks: #35]

tautschnig opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
More constant propagation

1) Within goto-symex, propagation of nondet_symbol_exprt is actually safe, as
they are instantiations of side_effect_expr_nondett and the value of these
instantiations cannot change.
2) This highlights that we also need to consider further expressions as possibly
being constant in is_constantt which we would otherwise simplify away when they
are constant (but cannot any longer do so when nondet_symbol_exprt are contained
in there).
3) Constant propagation now yields different right-hand sides in a test, and
makes another test trivial.
4) goto-symex applies filters on what is considered "constant" to keep the size
of expressions contained as (repeated) simplification of large expressions is
expensive. We may revisit this in future as we improve the simplifier.

tautschnig commented:
@smowton I'd be curious whether this can do some of what otherwise might only be achievable via field sensitivity for Java.

smowton commented:
Propagating a nondet symbol? I don't see how, could you elaborate what you think it might do?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3714d2507bda17b1d1924e3bee3d8c3f018af4dd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95553013

tautschnig commented:
> Propagating a nondet symbol? I don't see how, could you elaborate what you think it might do?

A nondet symbol is a constant, we just don't know its value. The gain is that it does not interfere with structs that are otherwise constants (those of the kind that we know their value), see also the regression test that's included.

kroening commented:
We should probably rename is_constantt to something like "forward_substitution".

kroening commented:
Specifically on Java: in principle, Java objects should get zero initialised -- this should enable the propagation without this PR. It may be worth checking whether that actually works.

smowton commented:
@kroening they do get zero-init'd; this would catch a case such as `otherwise_const_thing->field = nondet_value;`. Because various arithmetic operations on constants are themselves considered constants quite a lot would get propagated -- I think the objects would only stop getting propagated once there's a phi node, as `if_exprt` does not inherit constant-ness from its arguments.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f241b8be9d694ed2749f56b1e9eb7d19c65e023b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98502213
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

smowton commented:
My inclination is this is a bad thing (propagating not just `nondet_symbol#1` but also a big ball of non-constant arithmetic containing `nondet_symbol#1`) and will be rendered redundant by field sensitivity

tautschnig commented:
> [...] and will be rendered redundant by field sensitivity

I would have liked to believe this, but in fact these are very much complementary. On one of our verification tasks field sensitivity works as expected, but in order to handle a call to `memset` we _also_ need a pointer offset to propagate (neither the changes from this PR alone nor field sensitivity alone suffice), which is computed from a sequence of pointer arithmetic. The simplifier is then able to figure out that all the nondets do not matter for the pointer offset, and yields a much-needed constant.

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -153,6 +153,20 @@ class goto_symex_is_constantt : public is_constantt
+    else if(expr.id() == ID_if)
+    {
+      const if_exprt &if_expr = to_if_expr(expr);
+      if(
```
Seems oddly specific, add a comment saying why this particular restriction?

tautschnig replied:
Comment added.

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -153,6 +153,20 @@ class goto_symex_is_constantt : public is_constantt
+      return true;
+    else if(expr.id() == ID_symbol)
+    {
+      const irep_idt &obj_identifier = to_ssa_expr(expr).get_object_name();
```
Similarly could use a comment: why this policy for guards only?

peterschrammel replied:
Suggest to take the guard prefix from a single place of definition.

tautschnig replied:
Comment added and a refactoring commit added that makes all users use a single definition.

smowton reviewed (approved):
Fair enough, this is easy to revert if it has bad side-effects, so let's try it

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3a747ed82f5a9cb4df9316328ae9986e35865f89).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99236122

tautschnig commented:
I'm quite keen to merge this, but I am worried that it produces different test results: @allredj @romainbrenguier could you please set up some test to check that merging this PR won't break TG?

romainbrenguier commented:
@tautschnig 
> @romainbrenguier could you please set up some test to check that merging this PR won't break TG?

I've made the PR, I will update you when we get the results.

tautschnig commented:
Marking as do-not-merge as there appear to be TG failures that need investigation.

romainbrenguier reviewed (changes requested):
I tested this branch on the following example:
```
public class andNot_Bug {
    public static void main(int capacity) {
        if(capacity < 1)
            return;
        boolean b1 = true;
        boolean b2[] = new boolean[capacity];
        b2[0] = false;

        for (int i = 0; i < b2.length; i++) {
            b1 = b1 && !b2[i];
        }
    }
}
```
and jbmc aborts with an invariant failure. This needs to be addressed before it can be merged.

tautschnig commented:
Just noting that this test also fails with current develop (and would likely be fixed by #2068). We'll work on narrowing this down a bit further.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 794d0e93387eca1df573405d56519c372f5a0ce1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99337131

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fe3aad48855247ba0d0484aad78565334077602b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99637451

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f5e5d3e38b23279183bb2e0b4a9eb7220d98254d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99708343

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3600644dd04ad3d3cdb3597a8d413370a2c9b66a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100086630

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b9cd724d165bc1b36434df719d890709344c005d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100985691

romainbrenguier commented:
I updated the manual bump and it failed but less badly than before. @tautschnig maybe you can rebase this PR, it will make it easier to see if there is something wrong.

tautschnig commented:
@romainbrenguier Rebased - thank you very much for your help, please let me know if there is anything I could do to make debugging any easier for you!

tautschnig commented:
@romainbrenguier Updated again as one test now "failed," because it had become trivial. Checked in with @smowton about that and it seems ok for the test to be trivial.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bfb3fa4f26b885ce8fb91f85c2d68de028764752).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102836965

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 1dfd196522d60d7b27528bb419f3b312d6e04b1b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103082443
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

romainbrenguier commented:
Sorry about making this wait. The problems I see now seem to be related to strings. In this example:
```java
public class Test {
    public String foo() {
        final int i = 10;
        final String retval = Integer.toHexString(i);
        assert(false);
        return retval; 
    }
}
```
running `jbmc Test.class --function Test.foo --trace` used to give a trace in which we had:
```
State 30 function java.lang.Integer.toHexString() thread 0
----------------------------------------------------
  dynamic_object1={ 'a' } ({ 00000000 01100001 })
```

but with these changes, it is replaced by:
```
State 30 function java.lang.Integer.toHexString() thread 0
----------------------------------------------------
  dynamic_object1=dynamic_object1#2 (?)
```

tautschnig committed:
Add arrayst::get to return RHS of unbounded arrays

In order to generate a trace value for an unbounded LHS we may need to
look at its RHS as (string) refinement may have constructed values for
the RHS.

tautschnig commented:
@romainbrenguier I have now tracked down the problem and have added a second commit that addresses it, including the test that you provided. Would you be able to take another look that fully addresses the problem in your tests? Thanks!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4d0f07367c2d7c1c752b196226666e6fef5546c4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104052645

romainbrenguier commented:
I don't remember if I sent a similar example before, but here is a test for `jbmc` which was passing before, but fails on this branch:
```java
class BitSet {
    boolean bits[];
    public BitSet()
    {
        int size = org.cprover.CProver.nondetInt();
        org.cprover.CProver.assume(size >= 0);
        bits = new boolean[size];
    }
    public void set(int i)
    {
        org.cprover.CProver.assume(i < bits.length);
        bits[i] = true;
    }
    public boolean get(int i)
    {
        if(i < bits.length)
            return bits[i];
        return false;
    }
    public void andNot(BitSet other)
    {
        for (int i = 0; i < bits.length; i++) {
            bits[i] = bits[i] && !other.get(i);
        }
    }
}

public class andNot_Pass {
    public static void main(String args[]) {
        BitSet b1 = new BitSet();
        b1.set(1);
        b1.set(2);
        BitSet b2 = new BitSet();
        b2.set(0);
        b2.set(2);
        b1.andNot(b2);
        
        assert(!b1.get(0));
    }
}
```

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 69012b3ec11873a129dd299f3b336ba5e4e57f42).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105668394

martin-cs reviewed (approved):
Seems reasonable.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ef4b9bb92797f45b1940992897a4e9d07443a22d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105797178

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5caef7417670014d11cbb7ba83570c6ded43c156).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105812275

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 90c4a756894fa1e17bbea30c926a3a07a0ce24f5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105987217
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bcb894d960f99d6403c403b951bb954ef798177e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107052342

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 77f93e971e61c710ddab5b15bbfbe446d04a585a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107328775
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0c848d6f332971ac5bf33407132077172f714389).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107800556

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 85c963cb9fc10df833b579fffd22456d45335667).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108226898

TGWDB commented:
Closing due to age. If you believe this is erroneous please reopen and update the PR.
